File: DesignPatterns/01_GOF23/DP_examples/CSharp/03_Behavioral/Interpreter/NonterminalExpression.cs
namespace Behavioral.Interpreter;

/// <summary>
/// 对运算符进行解释
/// </summary>
public abstract class OperationExpression(ArithmeticExpr exp1, ArithmeticExpr exp2) : ArithmeticExpr
{
    public abstract double Interpret();
    protected ArithmeticExpr exp1 = exp1, exp2 = exp2;
}
public class AdditionExpression(ArithmeticExpr exp1, ArithmeticExpr exp2) : OperationExpression(exp1, exp2)  
{
    public override double Interpret() => exp1.Interpret() + exp2.Interpret();
}
public class SubtractionExpression(ArithmeticExpr exp1, ArithmeticExpr exp2) : OperationExpression(exp1, exp2)  
{
    public override double Interpret() => exp1.Interpret() - exp2.Interpret();
}
public class MultiplicationExpression(ArithmeticExpr exp1, ArithmeticExpr exp2) : OperationExpression(exp1, exp2)  
{
    public override double Interpret() => exp1.Interpret() * exp2.Interpret();
}
public class DivisionExpression(ArithmeticExpr exp1, ArithmeticExpr exp2) : OperationExpression(exp1, exp2)  
{
    public override double Interpret() => exp1.Interpret() / exp2.Interpret();
}

File: DesignPatterns/01_GOF23/DP_examples/Python/03_Behavioral/Interpreter/NonterminalExpression.py
# 非终结符表达式 - 处理运算符
class OperationExpression:
    def __init__(self, exp1, exp2):
        self.exp1 = exp1
        self.exp2 = exp2
    
    def interpret(self):
        pass

class AdditionExpression(OperationExpression):
    def interpret(self):
        return self.exp1.interpret() + self.exp2.interpret()

class SubtractionExpression(OperationExpression):
    def interpret(self):
        return self.exp1.interpret() - self.exp2.interpret()

class MultiplicationExpression(OperationExpression):
    def interpret(self):
        return self.exp1.interpret() * self.exp2.interpret()

class DivisionExpression(OperationExpression):
    def interpret(self):
        return self.exp1.interpret() / self.exp2.interpret()